compiler/compiler_test.go

```diff
@@ -41,27 +43,336 @@ func Bfunc() int {
 	return varA+varB
 }
 `
+
 	files := []source{{"fileA.go", []byte(fileA)}, {"fileB.go", []byte(fileB)}}
 
-	compare(t, "foo", files, false)
-	compare(t, "foo", files, true)
+	compareOrder(t, files, false)
+	compareOrder(t, files, true)
 }
 
-func compare(t *testing.T, path string, sourceFiles []source, minify bool) {
-	outputNormal, err := compile(path, sourceFiles, minify)
-	if err != nil {
-		t.Fatal(err)
-	}
+func TestDeclSelection_KeepUnusedExportedMethods(t *testing.T) {
+	src := `
+		package main
+		type Foo struct {}
+		func (f Foo) Bar() {
+			println("bar")
+		}
+		func (f Foo) Baz() { // unused
+			println("baz")
+		}
+		func main() {
+			Foo{}.Bar()
+		}`
+
+	srcFiles := []source{{`main.go`, []byte(src)}}
+	sel := declSelection(t, srcFiles, nil)
+
+	sel.DeclCode.IsAlive(`^\s*Foo = \$newType`)
+	sel.DeclCode.IsAlive(`^\s*\$ptrType\(Foo\)\.prototype\.Bar`)
+	sel.DeclCode.IsAlive(`^\s*\$ptrType\(Foo\)\.prototype\.Baz`)
+}
+
+func TestDeclSelection_RemoveUnusedUnexportedMethods(t *testing.T) {
+	src := `
+		package main
+		type Foo struct {}
+		func (f Foo) Bar() {
+			println("bar")
+		}
+		func (f Foo) baz() { // unused
+			println("baz")
+		}
+		func main() {
+			Foo{}.Bar()
+		}`
+
+	srcFiles := []source{{`main.go`, []byte(src)}}
+	sel := declSelection(t, srcFiles, nil)
+
+	sel.DeclCode.IsAlive(`^\s*Foo = \$newType`)
+	sel.DeclCode.IsAlive(`^\s*\$ptrType\(Foo\)\.prototype\.Bar`)
+
+	sel.DeclCode.IsDead(`^\s*\$ptrType\(Foo\)\.prototype\.baz`)
+}
+
+func TestDeclSelection_KeepUnusedUnexportedMethodForInterface(t *testing.T) {
+	src := `
+		package main
+		type Foo struct {}
+		func (f Foo) Bar() {
+			println("foo")
+		}
+		func (f Foo) baz() {} // unused
+
+		type Foo2 struct {}
+		func (f Foo2) Bar() {
+			println("foo2")
+		}
+
+ 		type IFoo interface {
+			Bar()
+			baz()
+		}
+		func main() {
+			fs := []any{ Foo{}, Foo2{} }
+			for _, f := range fs {
+				if i, ok := f.(IFoo); ok {
+					i.Bar()
+				}
+			}
+		}`
+
+	srcFiles := []source{{`main.go`, []byte(src)}}
+	sel := declSelection(t, srcFiles, nil)
+
+	sel.DeclCode.IsAlive(`^\s*Foo = \$newType`)
+	sel.DeclCode.IsAlive(`^\s*\$ptrType\(Foo\)\.prototype\.Bar`)
+
+	// `baz` is used to duck-type (via method list) against IFoo
+	// but the method itself is not used so can be removed.
```

Comment:
Clarified phasing a little bit.

```suggestion
	// `baz` signature metadata is used to check a type assertion against IFoo,
	// but the method itself is never called, so it can be removed.
```